chrononaut/versioned.py

```diff
@@ -70,113 +80,124 @@ def versions(self, before=None, after=None, return_query=False):
-        query = self.__history_mapper__.class_.query.filter_by(
-            **{k: getattr(self, k) for k in prim_keys}
+        # Get the primary keys for this table
+        prim_keys = [
+            mapper.get_property_by_column(k).key for k in mapper.primary_key if k.key != "version"
+        ]
+        obj_key = {k: getattr(self, k) for k in prim_keys}
+
+        # Find all previous versions that have the same primary keys and table name as myself
+        query = activity.query.filter(
+            activity.key == obj_key, activity.table_name.__eq__(mapper.local_table.name)
         )
 
         # Filter additionally by date as needed
         if before is not None:
-            query = query.filter(self.__history_mapper__.class_.changed <= before)
+            query = query.filter(activity.changed <= before)
         if after is not None:
-            query = query.filter(self.__history_mapper__.class_.changed >= after)
+            query = query.filter(activity.changed >= after)
 
         # Order by the version
-        query = query.order_by(self.__history_mapper__.class_.version)
+        if ordered:
+            query = query.order_by(activity.version)
 
         if return_query:
             return query
         else:
-            return query.all()
+            return [chrononaut_snapshot_to_model(self, m) for m in query.all()]
 
-    def version_at(self, at):
+    def version_at(self, at, return_snapshot=False):
         """Fetch the history model at a specific time (or None)
 
         :param at: The DateTime at which to find the history record.
-        :return: A history model at the given point in time or the model itself if that is current.
+        :param return_snapshot: Return just the object snapshot dict instead of the model.
+        :return: The HistorySnapshot model representing the model at the given point in time.
         """
-        query = self.versions(after=at, return_query=True)
-        history_model = query.first()
+        query = self.versions(before=at, ordered=False, return_query=True)
+        activity = self.metadata._activity_cls
+        history_model = query.order_by(activity.version.desc()).first()
+
         if history_model is None:
-            return self
+            return None
         else:
-            return history_model
+            return (
+                history_model.data
+                if return_snapshot
+                else chrononaut_snapshot_to_model(self, history_model)
+            )
 
     def has_changed_since(self, since):
         """Check if there are any changes since a given time.
 
         :param since: The DateTime from which to find any history records
         :return: ``True`` if there have been any changes. ``False`` if not.
         """
-        return self.version_at(at=since) is not self
+        # TODO: this is ambiguous, what if there were 2 changes which cancel each other out?
```

Comment:
is it still TODO?

Reply:
The method works just like it did before, but the ambiguity was there in the old version (and so it's still there). Since it's backwards compatible we can keep it as is for this version, but moving forward perhaps we'd need two separate functions (`is_same_as_version` and `was_modified_since` to reflect actual modifications vs end result ambiguity).